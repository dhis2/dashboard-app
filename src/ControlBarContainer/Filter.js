import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { grey700 } from 'material-ui/styles/colors';
import TextField from 'material-ui/TextField';
import IconButton from 'material-ui/IconButton';
import IconClear from 'material-ui/svg-icons/content/clear';
import isEmpty from 'd2-utilizr/lib/isEmpty';

import * as fromReducers from '../reducers';

const KEYCODE_ESCAPE = 27;

const styles = {
    filterField: {
        fontSize: '14px',
        width: '200px',
        height: '30px',
    },
    filterFieldInput: {
        top: '-9px',
        left: '1px',
    },
    filterFieldUnderline: {
        bottom: '10px',
    },
    filterFieldUnderlineFocus: {
        borderColor: '#aaa',
        borderWidth: '1px',
    },
    clearButton: {
        width: '20px',
        height: '24px',
        padding: 0,
        top: '-5px',
        left: '-18px',
    },
    clearButtonIcon: {
        width: '16px',
        height: '16px',
    },
};

class Filter extends Component {
    constructor(props) {
        super(props);

        this.state = {
            value: fromReducers.fromFilter.DEFAULT_NAME,
        };

        this.setFilterValue = this.setFilterValue.bind(this);
        this.handleKeyUp = this.handleKeyUp.bind(this);
    }

    componentWillReceiveProps(nextProps) {
        this.setState({
            value: nextProps.name,
        });
    }

    setFilterValue(event) {
        event.preventDefault();

        this.props.onChangeName(event.target.value);
    }

    handleKeyUp(event) {
        if (event.keyCode === KEYCODE_ESCAPE) {
            this.props.onChangeName();
        }
    }

    render() {
        return (
            <TextField
                className="FilterField"
                value={this.state.value}
                onChange={this.setFilterValue}
                onKeyUp={this.handleKeyUp}
                hintText="Filter dashboards"
                style={styles.filterField}
                inputStyle={styles.filterFieldInput}
                hintStyle={styles.filterFieldHint}
                underlineStyle={styles.filterFieldUnderline}
                underlineFocusStyle={styles.filterFieldUnderlineFocus}
            />
        );
    }
}

Filter.propTypes = {
    name: PropTypes.string,
    onChangeName: PropTypes.func,
};

Filter.defaultProps = {
    name: '',
    onChangeName: Function.prototype,
};

const ClearButton = ({ name, onChangeName }) => {
    const disabled = isEmpty(name);

    return (
        <IconButton
            style={Object.assign({}, styles.clearButton, {
                opacity: disabled ? 0 : 1,
            })}
            iconStyle={styles.clearButtonIcon}
            onClick={() => onChangeName()}
            disabled={disabled}
        >
            <IconClear color={grey700} />
        </IconButton>
    );
};

ClearButton.propTypes = {
    name: PropTypes.string.isRequired,
    onChangeName: PropTypes.func.isRequired,
};

export default props => (
    <Fragment>
        <Filter {...props} />
        <ClearButton {...props} />
    </Fragment>
);
