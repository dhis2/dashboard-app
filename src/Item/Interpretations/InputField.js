import React, { Component } from 'react';
import TextField from 'd2-ui/lib/text-field/TextField';
import Button from 'd2-ui/lib/button/Button';
import { colors } from '../colors';

const style = {
    button: {
        height: '30px',
        width: '16.84px',
        color: colors.charcoalGrey,
        fontFamily: 'inherit',
        fontSize: '13px',
        lineHeight: '15px',
    },
    container: {
        marginBottom: '5px',
    },
    text: {
        fontSize: '14px',
        fontStretch: 'normal',
    },
    textField: {
        width: '80%',
        display: 'inline-block',
    },
};

class InputField extends Component {
    state = {
        newText: '',
    };

    updateNewText = newText => {
        this.setState({ newText });
    };

    onClick = () => {
        this.props.onPost(this.state.newText);
        this.setState({ newText: '' });
    };

    render() {
        return (
            <div style={style.container}>
                <div style={style.textField}>
                    <TextField
                        multiline
                        value={this.state.newText}
                        rows={1}
                        rowsMax={8}
                        fullWidth
                        style={style.text}
                        placeholder={this.props.placeholder}
                        onChange={this.updateNewText}
                    />
                </div>
                <Button style={style.button} onClick={this.onClick}>
                    {this.props.postText}
                </Button>
            </div>
        );
    }
}

export default InputField;
