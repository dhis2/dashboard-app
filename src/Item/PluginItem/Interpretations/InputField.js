import React, { Component } from 'react';
import TextField from 'd2-ui/lib/text-field/TextField';
import Button from 'd2-ui/lib/button/Button';
import { colors } from '../../../colors';

const style = {
    button: {
        height: '30px',
        width: '16.84px',
        top: '5px',
        left: '10px',
        color: colors.charcoalGrey,
        fontFamily: 'inherit',
        fontSize: '13px',
        lineHeight: '15px',
    },
    container: {
        marginBottom: '5px',
        width: '100%',
        display: 'inline-flex',
    },
    text: {
        fontSize: '14px',
        fontStretch: 'normal',
    },
    textField: {
        width: '80%',
        display: 'inline-block',
    },
};

class InputField extends Component {
    state = {
        newText: this.props.text || '',
    };

    updateNewText = newText => {
        this.setState({ newText });
    };

    onClick = () => {
        this.props.onSubmit(this.state.newText);
        this.setState({ newText: '' });
    };

    render() {
        return (
            <div style={style.container}>
                <div style={style.textField}>
                    <TextField
                        multiline
                        value={this.state.newText}
                        rows={1}
                        rowsMax={8}
                        fullWidth
                        style={style.text}
                        placeholder={this.props.placeholder}
                        onChange={this.updateNewText}
                    />
                </div>
                <Button
                    disabled={!this.state.newText.length}
                    style={style.button}
                    onClick={this.onClick}
                >
                    {this.props.postText}
                </Button>
            </div>
        );
    }
}

export default InputField;
