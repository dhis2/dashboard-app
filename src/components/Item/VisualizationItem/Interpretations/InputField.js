import React, { Component } from 'react';
import PropTypes from 'prop-types';
import TextField from 'd2-ui/lib/text-field/TextField';
import Button from 'd2-ui/lib/button/Button';
import { colors } from '../../../../modules/colors';

import { Editor as RichTextEditor } from '@dhis2/d2-ui-rich-text';
import MentionsWrapper from '@dhis2/d2-ui-mentions-wrapper';

const style = {
    button: {
        height: '30px',
        width: '16.84px',
        top: '10px',
        left: '3px',
        color: colors.charcoalGrey,
        fontFamily: 'inherit',
        fontSize: '13px',
        lineHeight: '15px',
    },
    container: {
        marginBottom: '5px',
        width: '100%',
        display: 'inline-flex',
    },
    text: {
        fontSize: '14px',
        fontStretch: 'normal',
    },
    textField: {
        width: '80%',
        display: 'inline-block',
    },
};

class InputField extends Component {
    state = {
        newText: this.props.text || '',
    };

    updateNewText = newText => {
        this.setState({ newText });
    };

    onClick = () => {
        this.props.onSubmit(this.state.newText);
        this.setState({ newText: '' });
    };

    render() {
        return (
            <div style={style.container}>
                <div style={style.textField}>
                    <MentionsWrapper
                        d2={this.context.d2}
                        onUserSelect={this.updateNewText}
                    >
                        <RichTextEditor onEdit={this.updateNewText}>
                            <TextField
                                multiline
                                value={this.state.newText}
                                rows={1}
                                rowsMax={8}
                                fullWidth
                                style={style.text}
                                placeholder={this.props.placeholder}
                                onChange={this.updateNewText}
                            />
                        </RichTextEditor>
                    </MentionsWrapper>
                </div>
                <Button
                    disabled={!this.state.newText.length}
                    style={style.button}
                    onClick={this.onClick}
                >
                    {this.props.postText}
                </Button>
            </div>
        );
    }
}

InputField.contextTypes = {
    d2: PropTypes.object,
};

InputField.defaultProps = {
    text: null,
    placeholder: null,
    postText: null,
    onSubmit: null,
};

InputField.propTypes = {
    text: PropTypes.string,
    placeholder: PropTypes.string,
    postText: PropTypes.string,
    onSubmit: PropTypes.func,
};

export default InputField;
