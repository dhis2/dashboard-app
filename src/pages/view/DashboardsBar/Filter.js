import React, { useState } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import i18n from '@dhis2/d2-i18n'
import cx from 'classnames'
import { colors, IconSearch16, IconSearch24 } from '@dhis2/ui'

import ClearButton from './ClearButton'

import { useWindowDimensions } from '../../../components/WindowDimensionsProvider'
import {
    acSetDashboardsFilter,
    acClearDashboardsFilter,
} from '../../../actions/dashboardsFilter'
import { sGetDashboardsFilter } from '../../../reducers/dashboardsFilter'
import { isSmallScreen } from '../../../modules/smallScreen'

import classes from './styles/Filter.module.css'

export const KEYCODE_ENTER = 13
export const KEYCODE_ESCAPE = 27

const Filter = ({
    clearDashboardsFilter,
    expanded,
    filterText,
    setDashboardsFilter,
    onKeypressEnter,
    onSearchClicked,
}) => {
    const [focusedClassName, setFocusedClassName] = useState('')
    const [inputFocused, setInputFocus] = useState(false)
    const { width } = useWindowDimensions()

    const setFilterValue = event => {
        event.preventDefault()
        setDashboardsFilter(event.target.value)
    }

    const onKeyUp = event => {
        switch (event.keyCode) {
            case KEYCODE_ENTER:
                onKeypressEnter()
                clearDashboardsFilter()
                break
            case KEYCODE_ESCAPE:
                clearDashboardsFilter()
                break
            default:
                break
        }
    }

    const onFocus = event => {
        event.preventDefault()
        setFocusedClassName(classes.focused)
    }

    const onBlur = event => {
        event.preventDefault()
        setFocusedClassName('')
    }

    const onFocusInput = input => {
        if (input && inputFocused && isSmallScreen(width)) {
            return input.focus()
        }
    }

    const activateSearchInput = () => {
        onSearchClicked()
        setInputFocus(true)
    }

    return (
        <span
            className={cx(
                classes.container,
                expanded ? classes.expanded : classes.collapsed
            )}
        >
            <button
                className={classes.searchButton}
                onClick={activateSearchInput}
            >
                <IconSearch24 color={colors.grey800} />
            </button>
            <div
                className={cx(classes.searchArea, `${focusedClassName}`)}
                onFocus={onFocus}
                onBlur={onBlur}
            >
                <div className={classes.searchIconContainer}>
                    <IconSearch16 color={colors.grey800} />
                </div>
                <input
                    className={classes.input}
                    type="text"
                    placeholder={i18n.t('Search for a dashboard')}
                    onChange={setFilterValue}
                    onKeyUp={onKeyUp}
                    value={filterText}
                    data-test="search-dashboard-input"
                    ref={onFocusInput}
                />
                {filterText && (
                    <div className={classes.clearButtonContainer}>
                        <ClearButton onClear={clearDashboardsFilter} />
                    </div>
                )}
            </div>
        </span>
    )
}

Filter.propTypes = {
    clearDashboardsFilter: PropTypes.func,
    expanded: PropTypes.bool,
    filterText: PropTypes.string,
    setDashboardsFilter: PropTypes.func,
    onKeypressEnter: PropTypes.func,
    onSearchClicked: PropTypes.func,
}

const mapStateToProps = state => ({
    filterText: sGetDashboardsFilter(state),
})

const mapDispatchToProps = {
    setDashboardsFilter: acSetDashboardsFilter,
    clearDashboardsFilter: acClearDashboardsFilter,
}

export default connect(mapStateToProps, mapDispatchToProps)(Filter)
