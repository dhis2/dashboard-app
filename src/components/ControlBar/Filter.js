import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import i18n from '@dhis2/d2-i18n';
import InputField from '@material-ui/core/Input';
import InputAdornment from '@material-ui/core/InputAdornment';
import SearchIcon from '@material-ui/icons/Search';
import { colors } from '@dhis2/ui-core';

import ClearButton from './ClearButton';
import { DEFAULT_STATE_DASHBOARDS_FILTER_NAME } from '../../reducers/dashboardsFilter';

export const KEYCODE_ENTER = 13;
export const KEYCODE_ESCAPE = 27;

const styles = {
    filterField: {
        fontSize: '14px',
        width: '200px',
        height: '30px',
        top: '-4px',
    },
    searchIcon: {
        color: colors.grey700,
        width: '20px',
        height: '20px',
    },
};

export class Filter extends Component {
    constructor(props) {
        super(props);

        this.state = {
            value: DEFAULT_STATE_DASHBOARDS_FILTER_NAME,
        };
    }

    componentWillReceiveProps(nextProps) {
        this.setState({
            value: nextProps.name,
        });
    }

    setFilterValue = event => {
        event.preventDefault();

        this.props.onChangeName(event.target.value);
    };

    handleKeyUp = event => {
        switch (event.keyCode) {
            case KEYCODE_ENTER:
                this.props.onKeypressEnter();
                break;
            case KEYCODE_ESCAPE:
                this.props.onChangeName();
                break;
            default:
                break;
        }
    };

    render() {
        const { classes, name, onChangeName } = this.props;

        const startAdornment = (
            <InputAdornment position="start">
                <SearchIcon className={classes.searchIcon} />
            </InputAdornment>
        );

        const endAdornment =
            name !== '' && name !== null ? (
                <InputAdornment position="end">
                    <ClearButton onClear={() => onChangeName()} />
                </InputAdornment>
            ) : null;

        return (
            <InputField
                className={classes.filterField}
                placeholder={i18n.t('Search for a dashboard')}
                startAdornment={startAdornment}
                endAdornment={endAdornment}
                value={this.state.value}
                onChange={this.setFilterValue}
                onKeyUp={this.handleKeyUp}
            />
        );
    }
}

Filter.propTypes = {
    name: PropTypes.string,
    onChangeName: PropTypes.func,
};

Filter.defaultProps = {
    name: '',
    onChangeName: Function.prototype,
};

export default withStyles(styles)(Filter);
