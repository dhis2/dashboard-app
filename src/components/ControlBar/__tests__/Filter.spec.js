import React from 'react';
import { shallow } from 'enzyme';
import InputField from '@material-ui/core/Input';
import { Filter, KEYCODE_ENTER, KEYCODE_ESCAPE } from '../Filter';

describe('Filter', () => {
    let props;
    let shallowFilter;
    const name = 'rainbow';
    const filter = () => {
        if (!shallowFilter) {
            shallowFilter = shallow(<Filter {...props} />);
        }
        return shallowFilter;
    };

    beforeEach(() => {
        props = {
            name,
            onChangeName: jest.fn(),
            classes: {},
        };
        shallowFilter = undefined;
    });

    it('matches the snapshot when name property is empty', () => {
        expect(filter()).toMatchSnapshot();
    });

    describe('when updated name property is provided', () => {
        it('renders a InputField with correct value property', () => {
            const filterWrapper = filter();
            filterWrapper.setProps({ name: 'rainbow' });

            expect(filterWrapper.find(InputField).props().value).toEqual(
                'rainbow'
            );
            expect(props.onChangeName).not.toHaveBeenCalled();
        });
    });

    describe('when InputField value is changed', () => {
        beforeEach(() => {
            props.onChangeName = jest.fn();
        });

        it('triggers onChangeName property with correct value', () => {
            const newValue = 'fluttershy';
            const e = {
                target: { value: newValue },
                preventDefault: jest.fn(),
            };
            const inputField = filter().find(InputField);
            inputField.simulate('change', e);
            expect(e.preventDefault).toHaveBeenCalledTimes(1);
            expect(props.onChangeName).toHaveBeenCalledTimes(1);
            expect(props.onChangeName).toHaveBeenCalledWith(newValue);
        });
    });

    describe('when key is pressed', () => {
        beforeEach(() => {
            props.onKeypressEnter = jest.fn();
            props.onChangeName = jest.fn();
        });

        it('triggers onKeypressEnter when key is ENTER', () => {
            filter()
                .find(InputField)
                .simulate('keyUp', { keyCode: KEYCODE_ENTER });

            expect(props.onKeypressEnter).toHaveBeenCalledTimes(1);
            expect(props.onChangeName).not.toHaveBeenCalled();
        });

        it('triggers onChangeName when key is ESCAPE', () => {
            filter()
                .find(InputField)
                .simulate('keyUp', { keyCode: KEYCODE_ESCAPE });

            expect(props.onChangeName).toHaveBeenCalledTimes(1);
            expect(props.onChangeName).toHaveBeenCalledWith();
            expect(props.onKeypressEnter).not.toHaveBeenCalled();
        });
    });
});
