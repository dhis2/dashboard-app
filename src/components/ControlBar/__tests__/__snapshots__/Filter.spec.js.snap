// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter matches the snapshot when name property is empty 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Filter
    classes={Object {}}
    name="rainbow"
    onChangeName={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "className": undefined,
      "endAdornment": <WithStyles(WithFormControlContext(InputAdornment))
        position="end"
      >
        <ClearButton
          onClear={[Function]}
        />
      </WithStyles(WithFormControlContext(InputAdornment))>,
      "onChange": [Function],
      "onKeyUp": [Function],
      "placeholder": "Search for a dashboard",
      "startAdornment": <WithStyles(WithFormControlContext(InputAdornment))
        position="start"
      >
        <pure(SearchIcon) />
      </WithStyles(WithFormControlContext(InputAdornment))>,
      "value": "",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "className": undefined,
        "endAdornment": <WithStyles(WithFormControlContext(InputAdornment))
          position="end"
        >
          <ClearButton
            onClear={[Function]}
          />
        </WithStyles(WithFormControlContext(InputAdornment))>,
        "onChange": [Function],
        "onKeyUp": [Function],
        "placeholder": "Search for a dashboard",
        "startAdornment": <WithStyles(WithFormControlContext(InputAdornment))
          position="start"
        >
          <pure(SearchIcon) />
        </WithStyles(WithFormControlContext(InputAdornment))>,
        "value": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
